src/devices/bus/nes/bootleg.cpp

```diff
@@ -920,29 +899,44 @@ void nes_btl_dn_device::write_h(offs_t offset, uint8_t data)
 
 /*-------------------------------------------------
 
- WHIRLWIND-2706
+ WHIRLWIND
 
- Games: Meikyuu Jiin Dababa (FDS conversion) and a few
- others
+ Games: Bubble Bobble (x2), Falsion, Meikyuu Jiin Dababa, ProWres
 
- This PCB maps PRG in 0x6000-0x7fff
+ Similar PCB variants for FDS conversions by Whirlwind Manu. They
+ map PRG in 0x6000-0x7fff. They vary slightly in the use/nonuse
+ of CHR RAM, switchable banks, and valid register locations.
 
  iNES: mapper 108
 
- In MESS: Supported.
+ In MAME: Supported.
 
  -------------------------------------------------*/
 
-void nes_whirl2706_device::write_h(offs_t offset, uint8_t data)
+uint8_t nes_whirlwind_device::read_m(offs_t offset)
 {
-	LOG_MMC(("whirl2706 write_h, offset: %04x, data: %02x\n", offset, data));
-	m_latch = data;
+	LOG_MMC(("whirlwind read_m, offset: %04x\n", offset));
+	return m_prg[((m_reg[0] & m_prg_mask) * 0x2000) + (offset & 0x1fff)];
 }
 
-uint8_t nes_whirl2706_device::read_m(offs_t offset)
+void nes_whirlwind_device::write_h(offs_t offset, uint8_t data)
 {
-	LOG_MMC(("whirl2706 read_m, offset: %04x\n", offset));
-	return m_prg[(m_latch * 0x2000 + (offset & 0x1fff)) & (m_prg_size - 1)];
+	LOG_MMC(("whirlwind write_h, offset: %04x, data: %02x\n", offset, data));
+
+	if (m_chr_source == CHRRAM)           // submappers 1,3
+	{
+		// Bubble Bobble register 0xf000-0xffff, others use 0x8000-0xffff
+		if (m_mirroring == PPU_MIRROR_VERT || offset >= 0x7000)
+			m_reg[0] = data;
+	}
+	else                                  // submappers 2,4
+	{
+		if (m_vrom_chunks == 2)       // ProWres
+			m_reg[1] = data;
+		else if (offset >= 0x6000)    // Bubble Bobble alt
+			m_reg[0] = m_reg[1] = data;
```

Comment:
This looks a lot like a hack for separating two related but not identical devices based on some coincidental characteristic.

Reply:
welcome to the world of NES mappers...

Reply:
That's because it is. Mapper 108 basically represents 4 PCB types from the same bootlegger that are slightly different. I debated about whether to merge them. Only thing is, the two you highlighted? Those were the same UNL-BB device already :)
You can differentiate them based on CHR ROM size (first conditional), but then the other game only has registers above 0xE000. The lack of this hack in the previous version for a supposedly more clever heuristic is the reason the graphics corrupted in bublbfdsb (which they still do in FCEUX, the origin of said hack).

For now I recommend the hack. NesDev will eventually split the mappers up (or at least they've noted they should). I can break this up properly later if the hack is that unacceptable, but I'll need to dig into the horrors of nes_ines.hxx. We weren't properly loading mapper 108 roms to the correct device previously through the -cart picker, so we'd need to disambiguate there.

Reply:
Note the old version:
`if (!(offset & 0x1000))`
was even worse, because it's the exact same attempt at a disambiguating hack of two devices but it's hidden behind what at a glance may be plausible hardware behavior. (Oh, and it doesn't work either.)

Reply:
yeah, this issue here is if you emulate the NES you're basically expected to be compatible with the older NES mapper definitions.  Unfortunately many of those are a mess, shoving things that don't belong together.  The emulator is considered at fault if it doesn't implement things like this however.

yeah, there are attempts to document them properly now, but they're barely gaining any traction, because we all know what the console scene is like, and even with support for the newer formats, where the mappers have been correctly identified as different, you're expected to keep the older formats working.

I don't think there's anything wrong with the PR as such, it's just the ugly side of NES emulation in general.

Reply:
If the iNES “mapper” represents multiple board types that can only be distinguished by coincidental characteristics like ROM size, that shouldn’t be hacked into the cartridge devices.  If it’s necessary for supporting loose software in iNES format, it should be confined to the code where it probes/loads iNES files.  The cartridge devices should represent real hardware configurations.

Reply:
I revisited this and made the four submappers separate devices under a shared parent class. I also added the appropriate code to the iNES loader. Unfortunately in practice it will almost certainly fail since proper NES 2.0 headers on obscure bootlegs are rare. C'est la vie.